planningcenter/topbar/modules/api_request.ts

```diff
@@ -0,0 +1,37 @@
+export type Fetch = typeof fetch;
+
+export const defaultFetch: Fetch = (url, options) => {
+  return fetch(url, { credentials: "include", mode: "cors", ...options });
+};
```

Comment:
I understand the sentiment of "the first and second commandments of TopBar". The aspirations of being able to work with any framework as we would with Rails feels great in the abstract... but it has yet to become a reality.

I wonder if there's a middle ground here? My idea would be that the `defaultFetch` can provide some out of the box affordance for Rails sessions, while remaining open to being overridden in the hypothetical future of a non-Rails application that uses some other method of API authentication.

```suggestion
export const defaultFetch: Fetch = (url, options) => {
  const railsCsrfToken: null | string = inferRailsCsrfToken()
  if (railsCsrfToken === null) {
    return fetch(url, { credentials: "include", mode: "cors", ...options });
  } else {
    return railsCsrfFetch(url, pseuedoCodeBuildUpTheOptions);
  }
};

// copied from https://github.com/ministrycentered/giving/pull/4482
function railsCsrfFetch(url, { headers, ...rest) {
   return fetch(url, {
     headers: { ...headers, "X-CSRF-Token": inferRailsCsrfToken() },
     ...rest,
     credentials: "include",
     mode: "cors",
   })
}

function inferRailsCsrfToken(): null | string {
  return window.Rails?.csrfToken?() || document.querySelector('meta[name="csrf-token"]')?.content || null
}
```

Seeing `credentials: "include"` and `mode: "core"` in both branches of this idea, it _seems_ like we already require some knowledge from Top Bar for what our infrastructure expects from a fetch request?

Reply:
You make a good point here. I wanted to stay blissfully unaware of Rails, but that's a pretty abstract goal.

FWIW on `credentials` and `mode`, my thinking in including them by default is that they would work in any host application that used session authentication to talk to the API. The CSRF token implementation detail is Rails specific, and that's why I stopped short of it.

@chantastic I'd love to hear your thoughts on this. If you feel ok about adding the CSRF inference, I do too.

Reply:
i think including a Rails-specific fetch function still aligns with the goals, provided that it's still injectable.

suffice to say, i'm not dogmatic about not having Rails specific implementations exported — as long as they aren't pre-baked into the implementation.

If it were me, I'd forgo `defaultFetch` and export only `railsCsrfFetch` and `inferRailsCsrfToken`.
Folks can use these by their explicit names or write a custom function product side.

I like the naming orientation of `railsCsrfFetch` as you could naturally have a `nodeCsrfFetch` and `universalCsrfFetch` — should those days ever come.

Reply:
@chantastic the latest revision doesn't export anything from the `api_request` module. It retains only the `configuredFetch` API. If there's an app that needs to do something special to configure their fetch, I don't expect that app will gain anything from Topbar exporting functions to infer CSRF tokens. 

`api_request.ts` exports `defaultFetch` only to other modules within Topbar, not as part of the final bundle.

Reply:
@danielma i gotcha. i'm mostly curious why we're abstracting over `railsCsrfFetch` with `defaultFetch`.

if the intent is to buy into a rails environment by default, my preference would be to see direct communication of that default in the naming.

`defaultFetch` feels like name we give the generic (non-CSRF) version, *after* this is a multi-platform solution.

I think I see two possible outcomes:

**1. Keep it pure** 
* use a `defaultFetch` (without CSRF) by default
* export (as part of the public API) `railsCsrfFetch`
* require composition app-side

this is probably the most precedented pattern and virtually eliminates a need for a breaking defaults change down the line

**2. Screw purity. We live in Rails**
* make the default Rails-dependent by default
* don't export anything. don't require app-side composition
* what and see if we ever need anything else

of course, there's a part of me that feels like — even when we can guarantee Rails — it's appropriate to provide the opt-out `fetch` with the default options you've included.

---

Question: is there a practical environment that we have today that could need a non-CSRF version for texting and what not.

In almost all cases, I feel like we've made it easy enough by including the Rails-ready function. Exporting it, and requiring app-side composition, doesn't feel like over-reaching. And it has the benefit of keeping the library logically decoupled while being Rails ready.

Just some thoughts.

**just incase i'm still being too opaque**
my inclination would be to export `railsCsrfFetch` as public Topbar API.
we'd remove the conditional and rely on non-CSRF logic branch as a default.
apps would then import and inject `railsCsrfFetch` app-side.

I'm open to that being a step too far.
However, I've seen app-side injection save us SO MANY TIMES.

But I'm also open to Option 2, shipping it, and peeling out the platform-specific default when needed.

Reply:
i'm sorry if that's only 60% cogent.
trying to wrap as many loose ends as i can before being gone 'til wednesday.

Reply:
@chantastic thanks for all the clarification. I'm sorry I didn't understand it the first time. With this new clarity, I can see you were saying the same thing in your last comment, and I just missed it. 

I'm on board with this design, as a mid-point between the initial version (require apps to provide their own Rails-friendly fetch method) and the secondary (include Rails-specific behavior by default).